cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/RemoteTaskExecutor.java

```diff
@@ -140,13 +142,31 @@ public byte[] runTask(RunnableTaskRequest runnableTaskRequest) throws Exception
             String.format("Received exception %s for %s", e.getMessage(), runnableTaskRequest.getClassName()));
         }
       }, retryStrategy, Retries.DEFAULT_PREDICATE);
+    } catch (ServiceException se) {
+      Exception ex = getTaskException(se);
+      //emit metrics with failed result
+      emitMetrics(startTime, false, runnableTaskRequest, getAttempts(ex));
+      throw ex;
     } catch (Exception e) {
       //emit metrics with failed result
       emitMetrics(startTime, false, runnableTaskRequest, getAttempts(e));
       throw e;
     }
   }
 
+  private Exception getTaskException(ServiceException e) {
+    if (!"application/json".equals(e.getContentType())) {
```

Comment:
I don't think it's unexpected overall. You can have a lot of various error codes with different content on any network call. Having application-level exception is a special case, even if it should be the most often one for workers. Also generally any content-type should be checked in the HTTP layer as it's specific to it.
Could you please:
 * Instead of content-type add jsonDetails string field to ServiceException. In RemoteClient check content-type and if it's json, set this field in service exception and not include it into the test, text should just be "Service X is not available (code)". When the content-type is not json, just go with the previous message "Service %s is not available: %s". Note that you can't parse json yet as it can be different fro different users
 * In RemoteTaskExecutor if the jsonDetails is filled, do parse and it's fine to error out, but it would be great to add more details, like status code and generally say that it's in scope of the service call.
 * If there is no json, just throw the original ServiceError out. May be it's some timeout or similar, it's better to have it and not "Unexpected content-type"
P.S. It's kind of weird that in RemoteClient for retryable error codes we throw out a non-retryable ServiceException. Is it expected?

Reply:
Thanks, fixed.

> P.S. It's kind of weird that in RemoteClient for retryable error codes we throw out a non-retryable ServiceException. Is it expected?

the retryable execute() API is [execute(HttpRequest, Idempotency)](https://github.com/cdapio/cdap/blob/21c5799b4053dcaa7d3694f2a74e24cfbfe3de69/cdap-common/src/main/java/io/cdap/cdap/common/internal/remote/RemoteClient.java#L149). ServiceException is retried in executeIdempont().
Callers that want their request to be retried should call this method.

Reply:
Yes, it's just unobvious and implied through layers that `ServiceException` is retryable for idempotent calls. We should make it more obvious, may be even by introducing an interface for such exceptions.

Reply:
Not in scope of this PR though, jsut something to think about

Reply:
Ack. Filed https://cdap.atlassian.net/browse/CDAP-20363